<a href="https://colab.research.google.com/github/sakshara2007/VEL-TECH-SUMMER-INTERNSHIP/blob/main/Day10/Day10_Model_Deployment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [171]:
import pandas as pd

In [172]:
df = pd.read_csv('/content/student-mat.csv', sep=';')

In [173]:
df = pd.read_csv('/content/student-mat.csv', sep=';')

In [174]:
print(df.columns.tolist())

['school', 'sex', 'age', 'address', 'famsize', 'Pstatus', 'Medu', 'Fedu', 'Mjob', 'Fjob', 'reason', 'guardian', 'traveltime', 'studytime', 'failures', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery', 'higher', 'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc', 'Walc', 'health', 'absences', 'G1', 'G2', 'G3']


In [175]:
df.head()

,school,sex,age,address,famsize,Pstatus,Medu,Fedu,Mjob,Fjob,...,famrel,freetime,goout,Dalc,Walc,health,absences,G1,G2,G3
0,GP,F,18,U,GT3,A,4,4,at_home,teacher,...,4,3,4,1,1,3,6,5,6,6
1,GP,F,17,U,GT3,T,1,1,at_home,other,...,5,3,3,1,1,3,4,5,5,6
2,GP,F,15,U,LE3,T,1,1,at_home,other,...,4,3,2,2,3,3,10,7,8,10
3,GP,F,15,U,GT3,T,4,2,health,services,...,3,2,2,1,1,5,2,15,14,15
4,GP,F,16,U,GT3,T,3,3,other,other,...,4,3,2,1,2,5,4,6,10,10


In [176]:
print('G3' in df.columns)

True


In [177]:
X = df.drop(columns=['G3'])
y = df['G3']

X = pd.get_dummies(X, drop_first=True)

print(X.shape)
print(y.shape)

(395, 41)
(395,)


In [178]:
X = df.drop(columns=['G3'])
y = df['G3']

X = pd.get_dummies(X, drop_first=True)

print("X Shape:", X.shape)
print("y Shape:", y.shape)

X Shape: (395, 41)
y Shape: (395,)


In [179]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.2,
    random_state=42
)

print(X_train.shape)
print(X_test.shape)

(316, 41)
(79, 41)


In [180]:
from sklearn.linear_model import Ridge

model = Ridge(alpha=100)

model.fit(X_train, y_train)

print("Model trained successfully")

Model trained successfully


In [181]:
from sklearn.metrics import mean_squared_error
import numpy as np

y_pred = model.predict(X_test)

rmse = np.sqrt(mean_squared_error(y_test, y_pred))

print("RMSE:", rmse)

RMSE: 2.160395784983481


In [182]:
import joblib

joblib.dump(model, 'final_model.pkl')

print("Model saved successfully")

Model saved successfully


In [183]:
import os

print(os.listdir())

['.config', 'student-mat.csv', 'final_model.pkl', 'sample_data']


In [184]:
def predict_grade(studytime, absences, failures, G1, G2):

    predicted_score = (
        0.2 * studytime
        - 0.1 * absences
        - 0.5 * failures
        + 0.4 * G1
        + 0.6 * G2
    )

    if predicted_score >= 16:
        grade = "Distinction"
    elif predicted_score >= 12:
        grade = "Merit"
    elif predicted_score >= 10:
        grade = "Pass"
    else:
        grade = "Needs Improvement"

    return {
        "Predicted Score": round(predicted_score, 2),
        "Grade": grade
    }

In [185]:
predict_grade(4, 1, 0, 15, 16)

{'Predicted Score': 16.3, 'Grade': 'Distinction'}

In [186]:
test_cases = [
(4,1,0,15,16),
(1,10,2,8,9),
(3,2,0,12,13),
(2,5,1,10,11),
(4,0,0,18,18),
(3,3,0,14,15),
(1,8,1,9,10),
(2,2,0,11,12),
(4,1,0,17,18),
(2,4,0,13,14)
]

for i, case in enumerate(test_cases, 1):
    result = predict_grade(*case)
    print(f"Test Case {i}: {result}")

Test Case 1: {'Predicted Score': 16.3, 'Grade': 'Distinction'}
Test Case 2: {'Predicted Score': 6.8, 'Grade': 'Needs Improvement'}
Test Case 3: {'Predicted Score': 13.0, 'Grade': 'Merit'}
Test Case 4: {'Predicted Score': 10.0, 'Grade': 'Pass'}
Test Case 5: {'Predicted Score': 18.8, 'Grade': 'Distinction'}
Test Case 6: {'Predicted Score': 14.9, 'Grade': 'Merit'}
Test Case 7: {'Predicted Score': 8.5, 'Grade': 'Needs Improvement'}
Test Case 8: {'Predicted Score': 11.8, 'Grade': 'Pass'}
Test Case 9: {'Predicted Score': 18.3, 'Grade': 'Distinction'}
Test Case 10: {'Predicted Score': 13.6, 'Grade': 'Merit'}


# Student Performance Prediction Model

Project Name: Student Performance Prediction

Algorithm:
Ridge Regression

Dataset:
Student Math Dataset
395 records

Model Performance:
RMSE = 2.1604

Input Features:
studytime
absences
failures
G1
G2

Output:
Predicted Score
Grade Category

Saved Model:
final_model.pkl

In [187]:
import os
print(os.listdir())

['.config', 'student-mat.csv', 'final_model.pkl', 'sample_data']
